### ISSUE TITLE
Cytoscape Automated Testing Suite for Biomedical Visualization

### ISSUE THREAD

**Issue by jingjingbic** [nrnb/GoogleSummerOfCode#233]
### Background
[Cytoscape](https://www.cytoscape.org/) is an invaluable open-source tool in the field of biomedical research, offering robust data visualization capabilities, particularly for molecular networks. Its ecosystem, enriched by over 370 specialized apps available in the App Store, facilitates diverse functionalities ranging from database access to novel analysis methods. However, the recent development of [Cytoscape Web](https://github.com/cytoscape/cytoscape-web), an online variant, has introduced complexities in maintaining the platform's integrity and quality across versions. As the ecosystem expands, the challenge of ensuring consistent and reliable product releases grows. Thus, the development of sophisticated testing tools becomes imperative to uphold the high standards of Cytoscape's offerings.

### Goal
The primary goal of this project is to create an automated testing framework tailored for Cytoscape, with a specific focus on scientific figure preparation—a critical use case for the software. This framework will utilize a collection of Cytoscape session files and [CX](https://www.cytoscape.org/cx/) documents, integral in generating figures for various scientific publications. It aims to verify whether the newest versions of both Cytoscape and Cytoscape Web can accurately reproduce these figures. A key aspect of this project involves ensuring that figures produced by Cytoscape Web are consistent with those generated by the desktop version. The framework will facilitate communication with Cytoscape Desktop through the [CyRest API ](https://apps.cytoscape.org/apps/cyrest)and automate tests on Cytoscape Web using renowned frameworks like [Playwright](https://playwright.dev/) or [Cypress](https://www.cypress.io/). A significant part of the testing process will involve comparing newly generated images against established 'gold standard' figures, employing image analysis tools ([OpenCV](https://opencv.org/), [ImageMagick](https://imagemagick.org/index.php)) and advanced language models (such as Gemini and GPT-4) to produce comprehensive test reports.

### Size and Length of Project

- 175 hours
- 12 weeks

### Skills

Javascript, TypeScript, Python, Cytoscape

### Public Repository

- https://github.com/cytoscape/cytoscape-web
- https://github.com/cytoscape/

### Potential Mentors

- Jing Chen
- Kei Ono
- Dylan Fong
- Chris Churas

**Comment by khanspers** [comment]
NRNB has been accepted as a mentoring organization for GSoC 2024. The contributor application period if March 28 – April 2. Here are some useful links:

[GSoC contributor guide](https://google.github.io/gsocguides/student/)
[NRNB project proposal template](https://docs.google.com/document/d/1Zi6L38CHEeq2aL6xzv0Ozhd_Y6D71W3yCBGHplmxr6k/edit?usp=sharing)
[Eligibility requirements](https://developers.google.com/open-source/gsoc/faq#what_are_the_eligibility_requirements_for_participation)
[Full program timeline](https://developers.google.com/open-source/gsoc/timeline)

**Comment by jingjingbic** [comment]
> Hello @khanspers @jingjingbic @d2fong @coleslaw481, I'm reaching out to clarify a point regarding this project.
> 
> I wanted to confirm my understanding regarding the 'gold standard' figures. From what I gather, these are pre existing visualizations that serves as benchmark for evaluating accuracy of visualizations produced by Cytoscape Web and Cytoscape desktop version. Is this correct, or are there additional considerations we need to take into account?
> 
> Your insights and guidance on this matter would be very valuable as we move forward with the project.
> 
> Thank you for your time and consideration :)

You are correct. Those are existing figures or some figures that can be generated from a CX2 file using Cytoscape Desktop. We need to compare them with the figures generated from the same CX2 file using Cytoscape Web.

**Comment by ayushkoli772** [comment]
Thank you @jingjingbic, for clarifying the approach. I've now gained a clear understanding of the project's scope and requirements.

To demonstrate my capability and contribute effectively to the project, I would appreciate it if you could assign me specific tasks aligned with the project's goals.

Additionally, I would like to inquire about any other communication channels or platforms where I can reach out to in the future for discussions regarding project progress and any challenges encountered.

Thank you once again for your guidance and support. I look forward to working together on this exciting project.

**Comment by jingjingbic** [comment]
> Thank you @jingjingbic, for clarifying the approach. I've now gained a clear understanding of the project's scope and requirements.
> 
> To demonstrate my capability and contribute effectively to the project, I would appreciate it if you could assign me specific tasks aligned with the project's goals.
> 
> Additionally, I would like to inquire about any other communication channels or platforms where I can reach out to in the future for discussions regarding project progress and any challenges encountered.
> 
> Thank you once again for your guidance and support. I look forward to working together on this exciting project.

You can start with writing a small test using playwright to test a certain UI output in Cytoscape Web. This [tutorial] (https://github.com/cytoscape/cytoscape-web/wiki/Tutorial-For-Cytoscape-Web) should help you setup your dev environment.

**Comment by maxkfranz** [comment]
@jingjingbic @d2fong The GSOC student should use tools like Pixelmatch to diff the images: https://github.com/mapbox/pixelmatch

They can use a loose threshold if the images are expected to differ just a little bit.

Seeing diffs visually in the report page would help, as a person would probably have to visually verify the results.  Seeing a diff like this would help so you don't have to play the spot-the-differences game: https://github.com/mapbox/pixelmatch?tab=readme-ov-file#example-output

The report should have a feature to see the test cases that have large diffs, and it would be useful to have a sorted list with the biggest diffs at the top.

**Comment by khanspers** [comment]
NRNB has been accepted as a mentoring organization for GSoC 2025. **The contributor application period is March 24 – April 8**. Here are some useful links:

[GSoC contributor guide](https://google.github.io/gsocguides/student/)
[NRNB project proposal template](https://docs.google.com/document/d/1Zi6L38CHEeq2aL6xzv0Ozhd_Y6D71W3yCBGHplmxr6k/edit?usp=sharing)
[Eligibility requirements](https://developers.google.com/open-source/gsoc/faq#what_are_the_eligibility_requirements_for_participation)
[Full program timeline](https://developers.google.com/open-source/gsoc/timeline)

**Comment by PalashChitnavis** [comment]
@jingjingbic @keiono @coleslaw481 @maxkfranz 
Is this issue still open for GSoC 2025? If so, I’d love to contribute! I have prior experience with Cytoscape.js and am eager to work on creating a Playwright test to compare results.

**Comment by NeilIsrani** [comment]
@khanspers 
I am keen on contributing to this project. After looking at the codebase, I find that I have the skills to jump right in. Are there any email IDs from the mentors you can share so that I can ask some specific questions and see whether GSoc contributors may still apply?

**Comment by khanspers** [comment]
@NeilIsrani : The contributor application period is open since March 24, see the full timeline here: https://developers.google.com/open-source/gsoc/timeline
Note that you can submit a draft proposal in the GSoC interface and later update it based on feedback you get from mentors.

**Comment by sanaa-sys** [comment]
Hi @khanspers 

I am Ayesha Ali, a recent software engineering graduate from Monash University. Currently I am working as a Software Develop at ConserveIT where I am working with Javascript(jQuery, Handlebars), React, and Typescript with some testing frameworks like Selenium and Playwright. I have experience working with Python and building interactive dashboards using data from large datasets. During my internship with CSIRO Energy, I analyzed energy market data and developed visual dashboards using Python, which strengthened my skills in data handling and user-focused design. I am also developed side projects with LLM and Python like customer churn prediction: https://github.com/sanaa-sys/Customer-Churn-Prediction. I am eager to contribute to this project as I am familiar with most of the tech stack used in the project.

Looking forward to contributing

**Comment by Vinu8468** [comment]
Hi @jingjingbic and mentors,

I’ve been following this issue and found the idea of building an automated testing framework for Cytoscape Web really interesting, especially the part about verifying that figures generated from the same data remain consistent across versions.

A little about me... my background is somewhat interdisciplinary. I did my undergraduate degree in Biotechnology and I’m currently pursuing a Master’s in Big Data Biology at IBAB. through my coursework I’ve worked with python and SQL, and I’ve been learning tools and practices like Git, Linux environments, database design, and machine learning too. Recently I’ve also started learning Node.js and React to get more comfortable with web development.

Because of my biology background, I’ve seen how important tools like Cytoscape are for visualizing biological networks and preparing figures for research work, which is why this project caught my attention.

I would be happy to start by contributing in smaller ways ...like helping write or test Playwright scripts for UI outputs or figure generation workflows in Cytoscape Web while I get more familiar with the codebase.

Looking forward to learning more about the project and contributing where I can.

**Comment by Rishi2600** [comment]
Hi @jingjingbic  @keiono  @d2fong  @coleslaw481 @khanspers Hope y'll are doing good. 

I'm interested in working on this automated testing suite for GSoC 2026.

I've submitted a proof-of-concept PR to `cytoscape/cytoscape-web` demonstrating 
the core pipeline:
🔗 https://github.com/cytoscape/cytoscape-web/pull/511

**What the POC implements:**
- Loads real CX2 networks from existing test fixtures into Cytoscape Web
- Waits for full render completion via layout spinner detection
- Captures screenshots scoped to `[data-testid="cyjs-renderer"]`
- Pixel-level baseline comparison pipeline (0.00% diff across runs)
- Determinism verification — same network renders identically across two page loads

**All 3 tests passing:**
```
✓ renders with-visual-style network and matches baseline (0.00% diff)
✓ canvas is non-empty after loading
✓ rendering is deterministic across two loads (0.00% diff)
```

I understand the full scope goes much deeper — OpenCV/ImageMagick comparison, 
CyREST integration for Desktop cross-validation, AI-powered semantic validation 
(GPT-4/Gemini), 20+ scientific session files, and CI/CD pipeline. I'm preparing 
my full proposal covering the 175-hour timeline.

Happy to discuss further!
